src/components/include/cbuf.h

```diff
@@ -405,241 +341,131 @@ __cbuf_alloc_meta_inconsistent(struct cbuf_alloc_desc *d, struct cbuf_meta *m)
  *
  * precondition:  size must be a power of 2 && >= PAGE_SIZE
  */
-static inline struct cbuf_alloc_desc *
-__cbufp_freelist_get(int size)
+static inline struct cbuf_meta *
+__cbuf_freelist_get(int size)
 {
+	struct cbuf_freelist *fl = PERCPU_GET(cbuf_alloc_freelists);
 	int order = ones(size-1) - PAGE_ORDER;
-	struct cbuf_alloc_desc *d;
-
+	struct cbuf_meta *m;
 	assert(pow2(size) && size >= PAGE_SIZE);
 	assert(order >= 0 && order < WORD_SIZE-PAGE_ORDER);
-	d = &cbufp_alloc_freelists[order];
-	assert(d->length == size);
-
-	return d;
+	m = &(fl->freelist_head[order]);
+	if (!CBUFM_GET_NEXT(m)) {
+		CBUFM_SET_NEXT(m, m);
+	}
+	return m;
 }
 
 static inline void *
-__cbuf_alloc(unsigned int sz, cbuf_t *cb, int tmem)
+cbuf_alloc(unsigned int sz, cbuf_t *cb, int tmem)
 {
 	void *ret;
-	struct cbuf_alloc_desc *d, *fl;
-	int cbid, len, already_used, mapped_in, flags;
-	struct cbuf_meta *cm;
+	int cbid, len, already_used, inconsistent, mapped_in, flags;
+	struct cbuf_meta *cm, *fl, old_head, new_head;
+	unsigned long long *target, *old, *update;
 	long cbidx;
-
-	if (tmem) {
-		fl = &cbuf_alloc_freelists;
-	} else {
-		/* need a size >= PAGE_ORDER, that is a power of 2 */
-		sz = nlepow2(round_up_to_page(sz));
-		fl = __cbufp_freelist_get(sz);
-	}
-	CBUF_TAKE();
+	sz = nlepow2(round_up_to_page(sz));
+	fl = __cbuf_freelist_get(sz);
 again:
-	d = FIRST_LIST(fl, next, prev);
-	if (unlikely(EMPTY_LIST(d, next, prev))) {
-		d    = __cbuf_alloc_slow(sz, &len, tmem);
-		assert(d);
-		ret  = d->addr;
-		cbid = d->cbid;
+	target = (unsigned long long *)(&fl->next_flag);
```

Comment:
Why are these `unsigned long long`s?  If 64 bits is required, then we might want to just use `u64_t`, but I am not quite following why 64 bits are needed.